stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
   }
 }
 
+// Elements in `outputs` can come from two sources:
+// a) Symbolic tensors produced by tensor ops, and added as trace graph nodes.
+// b) Concrete tensors produced by host code (e.g. Tensor(1.0)).
+func finalizeTraceFn(fnName: String,
+                     outputs: [CTensorHandle]) -> TraceContext {
+  guard let traceContext = _RuntimeConfig.traceState.context else {
+    fatalError("Not in tracing mode!.")
+  }
+  _RuntimeConfig.traceState = .notTracing
+  traceContext.finalize(traceeBasicName: fnName, outputs: outputs)
+  return traceContext
+}
+
+private extension TensorArrayProtocol {
+  // The returned handles are owned by the caller.
+  var cTensorHandles: [CTensorHandle] {
+    debugLog("Getting \(self._tensorHandleCount) C handles.")
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(
+      capacity: Int(self._tensorHandleCount))
+    debugLog("Unpacking handles into buffer.")
+    _unpackTensorHandles(into: buffer)
+    let status = TF_NewStatus()
+    debugLog("Copying buffer content to output handles.")
+    var output: [CTensorHandle] = []
+    for i in 0..<Int(_tensorHandleCount) {
+      let address = buffer.advanced(by: i)
+      let isConcrete = TFE_TensorHandleIsConcrete(address.pointee) != 0
+      debugLog("""
+                 Copying the \(i)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete).
+                 """)
+      let newHandle = TFE_TensorHandleCopySharingTensor(address.pointee,
```

Comment:
I tried this, but then the use of `\(newHandle)` in `debugLog()` would trigger warnings that this is still an optional value. Since we want to print the unwrapped pointer value, I did not make this change.